src/data/list/rotate.lean

```diff
@@ -105,4 +110,239 @@ by rw ← list.take_append_drop (n % list.length (a :: l)) (a :: l) at hl;
 
+lemma rotate_perm (l : list α) (n : ℕ) : l.rotate n ~ l :=
+begin
+  rw rotate_eq_rotate',
+  induction n with n hn generalizing l,
+  { simp },
+  { cases l with hd tl,
+    { simp },
+    { rw rotate'_cons_succ,
+      exact (hn _).trans (perm_append_singleton _ _) } }
+end
+
+@[simp] lemma nodup_rotate {l : list α} {n : ℕ} : nodup (l.rotate n) ↔ nodup l :=
+(rotate_perm l n).nodup_iff
+
+@[simp] lemma rotate_eq_nil_iff {l : list α} {n : ℕ} : l.rotate n = [] ↔ l = [] :=
+begin
+  induction n with n hn generalizing l,
+  { simp },
+  { cases l with hd tl,
+    { simp },
+    { simp [rotate_cons_succ, hn] } }
+end
+
+@[simp] lemma rotate_singleton (x : α) (n : ℕ) :
+  [x].rotate n = [x] :=
+begin
+  induction n with n hn,
+  { simp },
+  { rwa [rotate_cons_succ] }
+end
+
+lemma zip_with_rotate_distrib {α β γ : Type*} (f : α → β → γ) (l : list α) (l' : list β) (n : ℕ)
+  (h : l.length = l'.length) :
+  (zip_with f l l').rotate n = zip_with f (l.rotate n) (l'.rotate n) :=
+begin
+  cases l.length.zero_le.eq_or_lt with hl hl,
+  { simp [eq_nil_of_length_eq_zero hl.symm] },
+  { have : n = n % l.length + l.length * (n / l.length) := (nat.mod_add_div _ _).symm,
+    have hn : n % l.length < l.length := nat.mod_lt _ hl,
+    rw [this, ←rotate_rotate, ←rotate_rotate, ←rotate_rotate, rotate_length_mul',
+        rotate_length_mul', rotate_length_mul',
+        rotate_eq_take_append_drop, rotate_eq_take_append_drop,
+        rotate_eq_take_append_drop, zip_with_append, ←zip_with_distrib_drop,
+        ←zip_with_distrib_take],
+    all_goals { simp [←h, hn.le] } }
+end
+
+local attribute [simp] rotate_cons_succ
+
+@[simp] lemma zip_with_rotate_one {β : Type*} (f : α → α → β) (x y : α) (l : list α) :
+  zip_with f (x :: y :: l) ((x :: y :: l).rotate 1) =
+  f x y :: zip_with f (y :: l) (l ++ [x]) :=
+by simp
+
+lemma nth_le_rotate_one (l : list α) (k : ℕ) (hk : k < (l.rotate 1).length) :
+  (l.rotate 1).nth_le k hk = l.nth_le ((k + 1) % l.length)
+    (by { cases l, { simpa using hk }, { refine nat.mod_lt _ _, simp }}) :=
+begin
+  cases l with hd tl,
+  { simp },
+  { have : k ≤ tl.length,
+      { refine nat.le_of_lt_succ _,
+        simpa using hk },
+    rcases this.eq_or_lt with rfl|hk',
+    { simp [nth_le_append_right (le_refl _)] },
+    { simpa [nth_le_append _ hk', length_cons, nat.mod_eq_of_lt (nat.succ_lt_succ hk')] } }
+end
+
+lemma nth_le_rotate (l : list α) (n k : ℕ) (hk : k < (l.rotate n).length) :
+  (l.rotate n).nth_le k hk = l.nth_le ((k + n) % l.length)
+    (by { cases l, { simpa using hk }, { refine nat.mod_lt _ _, simp }}) :=
+begin
+  induction n with n hn generalizing l k,
+  { have hk' : k < l.length := by simpa using hk,
+    simp [nat.mod_eq_of_lt hk'] },
+  { simp [nat.succ_eq_add_one, ←rotate_rotate, nth_le_rotate_one, hn l, add_comm, add_left_comm] }
+end
+
+lemma rotate_eq_iff {l l' : list α} {n : ℕ} :
+  l.rotate n = l' ↔ l = l'.rotate (l'.length - n % l'.length) :=
+begin
```

Comment:
I think this lemma would help shorten the proof:
```lean
lemma rotate_injective (n : ℕ) : function.injective (λ l : list α, l.rotate n) :=
begin
  rintros l l' (h : l.rotate n = l'.rotate n),
  have hle : l.length = l'.length := (l.length_rotate n).symm.trans (h.symm ▸ l'.length_rotate n),
  rw [rotate_eq_take_append_drop', rotate_eq_take_append_drop'] at h,
  obtain ⟨ht, hd⟩ := append_inj h _,
  { rw [←l.take_append_drop (n % l.length), ←l'.take_append_drop (n % l'.length), ht, hd], },
  { rw [length_drop, length_drop, hle], }
end

-- possibly easier to find in doc-gen, otherwise not that useful.
lemma rotate_eq_rotate {l l' : list α} {n : ℕ} :
  l.rotate n = l'.rotate n ↔ l = l' :=
(rotate_injective n).eq_iff
```